kuri_camera/src/ros_publisher.cpp

```diff
@@ -43,44 +42,42 @@ struct KuriCameraROSPublisher {
 
   void listen_publish_loop() {
     ros::Rate sub_poll(1);
+    ros::Rate fps(20);
+    cv::Mat image;
+    ros::Time timeForMsg;
+    //cv::Mat imageMat;
     while (ros::ok()) {
       ros::spinOnce();
-      if (publisher.getNumSubscribers() == 0) {
-        ROS_INFO("No subs");
+      fps.sleep();
+      if (imagePublisher.getNumSubscribers() == 0) {
+        ROS_INFO("Waiting for subscribers");
         sub_poll.sleep();
-        //continue;
-      }
-      ROS_INFO("Heard sub");
-      // Someone is subscribed! Get a frame
-      {
-        std::unique_lock<std::mutex> lock(m);
-        frame_processed = false;
-        got_frame.wait(lock, [this]() { return frame_processed; });
-        ROS_ERROR("LOOP SEES PROCESSED");
+        continue;
       }
 
-
+      std::unique_lock<std::mutex> imageLock(imageMutex);
+      if (hasNewImage) {
+        hasNewImage = false;
+        image = cv::imdecode(imageMat, CV_LOAD_IMAGE_UNCHANGED);
+        timeForMsg = recvTime;
+        imageLock.unlock();
+        sensor_msgs::CompressedImagePtr msg = cv_bridge::CvImage(std_msgs::Header(), "bgr8", image).toCompressedImageMsg();
+        msg->header.stamp = timeForMsg;
+        msg->header.frame_id = "upward_looking_camera";
+        imagePublisher.publish(msg);
+      } else {
+        imageLock.unlock();
+      }
     }
   }
 
   void data_callback(uint8_t *buffer, uint32_t size) {
-    std::unique_lock<std::mutex> lock(m);
-    std::cout << size << std::endl;
-    std::cout << (void*)buffer << std::endl;
-    std::cout << buffer[0] << std::endl;
-
-    if (size < 1000) {
-      return;
-    }
-    cv::Mat image_nv12(640, 480, CV_8UC1, buffer);
-    cv::Mat image_bgr;
-    cv::cvtColor(image_nv12, image_bgr, CV_YUV2BGR_YV12);
-    sensor_msgs::ImagePtr msg = cv_bridge::CvImage(std_msgs::Header(), "bgr8", image_bgr).toImageMsg();
-    publisher.publish(msg, camera_info);
-    frame_processed = true;
-    lock.unlock();
-    got_frame.notify_all();
-    ROS_ERROR("CALLBACK RETURNING");
+    if (size < 1000) return;
+    std::unique_lock<std::mutex> imageLock(imageMutex);
```

Comment:
It's good to move the processing out of the callback, but it does introduce a small race condition; the moment the lock is released, either the main loop will grab the lock to process and publish the frame, or the next callback will grab the lock and overwrite the frame. This might be exacerbated by the low poll rate for the frame processing.

Reply:
I would argue that this is a feature, not a bug. We want to publish the most recent image, not all the images (especially since the MJPEG format does not have temporal compression). That way, when the image processor is ready for another image, it will grab the most recent image, but it won't create a backlog due to the fact that the processing is slower than the camera stream. I don't think the worst case you mentioned (where the callback never gives up the lock) will ever happen, because when the processor is waiting for the lock it will block until the lock is available, so as soon as the callback frees it the processor will grab it.

I agree the low FPS is a problem -- I increased that.